### PR TITLE
AYON: Use folder path as unique identifier

### PR DESCRIPTION
**Pull request by iLLiCiTiT**
## Changelog Description
Use folder path instead of asset name as unique identifier, with OpenPype compatibility.

## Additional info
Goal of this PR is to support duplicated folder names in a project for AYON mode. How it works? All places that have asset name as "unique identifier" and not as "name" value are using folder path now. Some places do/did use the value for other purposes, all of those places must use different approach.

### What it affects
IMPORTANT FOR DEVELOPERS
- `get_current_asset_name` returns path in AYON mode
- `os.environ["AVALON_ASSET"]` and `legacy_io.Session["AVALON_ASSET"]` contains path in AYON mode
- during publishing `context.data["asset"]` and `instance.data["asset"]` contains path in AYON mode
- all pipeline functions that expect asset name as unique identifier (e.g. `get_template_data_with_names`) expect path instead of name in AYON mode
- creators (except legacy creators) should use `"folderPath"` instead of `"asset"`
- `openpype.client` functions expecting asset name are can handle folder path in AYON mode > the functions still can find by name
    - `get_assets`
    - `get_archived_assets`
    - `get_asset_by_name`
    - `get_last_version_by_subset_name`

### TODOs
- folder path will start with slash in future version of AYON server, so that must be changed acordingly
- editorial publishing was not modified at all (I'm not capable of doing that)
    - [hiero](https://github.com/ynput/OpenPype/pull/5829)
    - flame
    - resolve
    - traypublisher
- revisit Harmony `openpype/hosts/harmony/plugins/load/load_template.py` line 55 > There is passed asset name but I don't know what for
- revisit `ExtractUnrealSkeletalMeshFbx` plugin in Maya which is using asset name and not sure for what
- 3dsmax has some mismatched data in `CollectRender` plugin (different PR to resolve [In progress])
- houdini `CreateHDA` has one occation where it cannot work with current implementation
- unreal has at multiple places mismatch with OpenPype asset name and Unreal asset name
- host integrations without publisher (using legacy create tool) are not converted
    - blender
    - harmony

### Folder path have to start with slash in on AYON server
- backend PR https://github.com/ynput/ayon-backend/pull/56

### Allow to create duplicated names on WebUI
- frontend branch [enhancement/allow_duplicate_folders](https://github.com/ynput/ayon-frontend/tree/enhancement/allow_duplicate_folders)

### Testing
1. Have AYON project which allows duplicated names.
    - To allow this you have to modify database on running AYON docker stack
    a) Go to your ayon-docker repository.
    b) Run `./manage.ps1 dbshell` or `make dbshell`
    c) Execute `DROP INDEX project_{project name}.folder_unique_name;` where `{project name}` is replaced by name of your project
2. Create duplicated names in the project hierarchy
3. Run DCCs on a folder with duplicated name

- Try to use all creators and publish them, and validate the output ended up at correct places.
    - Make sure the template of `{asset}` or `{folder[name]}` is NOT filled with path.
- Validate all tools > visual aspects.
- Report any issues (check if the issue was reported first).

- open scene with created instances before this PR -> what happens in Publisher

### Hosts
- [x] AfterEffects @kalisp 
- [x] Blender @simonebarbieri - needs revisit after https://github.com/ynput/OpenPype/pull/5910
- [x] Celaction
- [ ] <strike>Flame @jakubjezek001 </strike> Won't be converted at this point.
- [x] Fusion @jakubjezek001 
- [ ] <strike>Harmony @kalisp </strike> Will be dealt with in own addon.
- [x] Hiero @jakubjezek001 https://github.com/ynput/OpenPype/pull/5829
- [x] Houdini @MustafaJafar 
- [x] 3dsMax @moonyuet 
- [x] Maya @antirotor 
- [x] Nuke @mkolar
- [x] Photoshop @kalisp
- [x] Resolve @jakubjezek001  https://github.com/ynput/OpenPype/pull/5892
- [x] Substance Painter @LiborBatek 
- [x] TrayPublisher @iLLiCiTiT
- [x] TrayPublisher - editorial @jakubjezek001 https://github.com/ynput/OpenPype/pull/5873
- [x] TVPaint @iLLiCiTiT 
- [x] Unreal @simonebarbieri

### CODE REVIEW

File: openpype/hosts/houdini/plugins/create/create_workfile.py
Comment (BigRoy) [inline openpype/hosts/houdini/plugins/create/create_workfile.py:39]: Wouldn't this technically fail when opening a workfile built with OpenPype in a project converted to AYON now opened in AYON.

The currently existing instance does **not yet** have `folderPath` defined even though `AYON_SERVER_ENABLED` is True?

Also, I wonder whether it makes sense putting this somewhere in a library for now so it can be re-used:
```python
def get_instance_asset(instance):
    if AYON_SERVER_ENABLED and "folderPath" in instance:
        return instance["folderPath"]
    return instance.get("asset")
```

Comment (iLLiCiTiT) [reply openpype/hosts/houdini/plugins/create/create_workfile.py]: I didn't realized I'm working with already created object of `CreatedInstance`.

File: openpype/client/server/entities.py
Comment (Minkiu) [inline openpype/client/server/entities.py:229]: It would be assumed that this would never reach if line `222` was true (so we do have `asset_names`), right?

File: openpype/hosts/hiero/plugins/publish/precollect_workfile.py
Comment (Minkiu) [inline openpype/hosts/hiero/plugins/publish/precollect_workfile.py:23]: Isn't in AYON that we want it with the slashes?

Comment (iLLiCiTiT) [reply openpype/hosts/hiero/plugins/publish/precollect_workfile.py:23]: Here the variable should represent name, not path.

File: openpype/client/server/entities.py
Comment (iLLiCiTiT) [reply openpype/client/server/entities.py:229]: Yep, you're right.

File: openpype/client/server/entities.py
Comment (iLLiCiTiT) [inline openpype/client/server/entities.py:234]:
```suggestion
    for name in asset_names:
        if "/" in name:
            folder_paths.add(name)
        else:
            new_asset_names.add(name)
```

File: openpype/hosts/resolve/plugins/publish/precollect_workfile.py
Comment (Minkiu) [inline openpype/hosts/resolve/plugins/publish/precollect_workfile.py:18]: We are not checking if `AYON_SERVER_ENABLED`  ?

Comment (iLLiCiTiT) [reply openpype/hosts/resolve/plugins/publish/precollect_workfile.py:18]: Well, it does not hurt in OpenPype ("sh0010" will stay "sh0010"), but I will add it.

File: openpype/hosts/tvpaint/plugins/create/create_review.py
Comment (Minkiu) [inline openpype/hosts/tvpaint/plugins/create/create_review.py:39]: This doesn't match the other filed where you do the `elif`

File: openpype/hosts/tvpaint/plugins/create/create_workfile.py
Comment (Minkiu) [inline openpype/hosts/tvpaint/plugins/create/create_workfile.py:35]: Doesn't match the other files where you do `elif`

**Comment by Minkiu** [comment]
Haven't tested, code looks fine, I had a overall suggestion, for most instances where we do the following:
```py
if AYON_SERVER_ENABLED:
    new_asset_name = instance.get("folderPath")
else:
    new_asset_name = instance.get("asset")
```

It might be cleaner if at the top of the class (or earlier) we do:
```py
asset_name_key = "folderPath" if AYON_SERVER_ENABLED "asset"
# And then later on, isntead of doing the `if..else` we use that key:
# getting it
new_asset_name = instance.get(asset_name_key)
# setting it
data[asset_name_key] = asset_name
```

Or it could be a constant, not sure, but I think it would reduce the amount of `if..else`

**Comment by jakubjezek001** [comment]
> I do agree it would probably reduce the code, but my thoughts were that we should visibly distinct/point out: This does not work the same way in AYON and OpenPype.

It seems to be clear enough from `asset_name_key = "folderPath" if AYON_SERVER_ENABLED "asset"`

**Comment by jakubjezek001** [comment]
> Have AYON project which allows duplicated names.
> To allow this you have to modify database on running AYON docker stack
> a) Go to your ayon-docker repository.
> b) Run ./manage.ps1 dbshell or make dbshell
> c) Execute DROP INDEX project_{project name}.folder_unique_name; where {project name} is replaced by name of your project
> Create duplicated names in the project hierarchy

The testing notes are not working as expected. It seems like the ablity is blocked by frontend. 
image

**Comment by martastain** [comment]
> > Have AYON project which allows duplicated names.
> > To allow this you have to modify database on running AYON docker stack
> > a) Go to your ayon-docker repository.
> > b) Run ./manage.ps1 dbshell or make dbshell
> > c) Execute DROP INDEX project_{project name}.folder_unique_name; where {project name} is replaced by name of your project
> > Create duplicated names in the project hierarchy
> 
> The testing notes are not working as expected. It seems like the ablity is blocked by frontend. image

Thanks, it is indeed blocked by the frontend before it sends the request. This restriction is now disabled in `experimental` image (but dropping the index from the database is still necessary).

**Review by jakubjezek001 — APPROVED** [review]
Launcher works:
image

Workfile works:
image

Louder works:
image

**Review by jakubjezek001 — CHANGES_REQUESTED** [review]
So after thorough investigation I have found one thing which might be issue in integrator.py 

I have loaded previously published plate from Traypublisher into Nuke and path was `SyncSketchTesting/shots/sq001//shots/sq001/sh010/publish/plate/plateMain/v001/SST_/shots/sq001/sh010_plateMain_v001.####.exr`

Here is the browser screegrub 
image

**Comment by iLLiCiTiT** [comment]
> So after thorough investigation I have found one thing which might be issue in integrator.py

Should be fixed.

**Comment by BigRoy** [comment]
> host integrations without publisher (using legacy create tool) are not converted
> 
>     * blender
> 
>     * harmony

Note that **Resolve** and **Hiero** also use the old publisher.

> ### What it affects
> 
> IMPORTANT FOR DEVELOPERS

I know it's still draft - but in the end it'd be good to tweak the PR description to be more informative about what a developer should do.

Can we get some notes added per statement there **what** in particular it affects and how to go about it now. For example `get_current_asset_name`. Is that **deprecated** and should it be unused? what's the replacement? etc.

**Review by mkolar — CHANGES_REQUESTED** [review]
Trying to change folder in a new publisher doesn't work if choosing a same folder name in a different path. 
Changing from `sq010/sh010` to `sq010/sh020` works just fine
changing from `sq010/sh010` to `sq020/sh010` doesn't work

**Review by mkolar — CHANGES_REQUESTED** [review]
Maya now works for most things apart from validating publishing in context. It used `context.data[AssetEntity"]` which is still collected in global plugin `openpype/plugins/publish/collect_context_entities.py` to reflect the name of asset/folder without a path. 

Easiest way to test is to simply open maya and try to publish a workfile. It will throw validation error here `openpype/hosts/maya/plugins/publish/validate_instance_in_context.py` because it compares full path vs plain folder name.

**Comment by antirotor** [comment]
in maya during startup - this is probably related:

```
# Warning: EventCallback : Failed to execute event callback < EventCallback - on_new > C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\hosts\maya\api\pipeline.py
# Traceback (most recent call last):
#   File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\lib\events.py", line 163, in process_event
#     callback()
#   File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\hosts\maya\api\pipeline.py", line 631, in on_new
#     lib.set_context_settings()
#   File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\hosts\maya\api\lib.py", line 2453, in set_context_settings
#     reset_scene_resolution()
#   File "C:\Users\annat\Documents\Projects\OpenPype\sources\ayon-desktop\openpype\hosts\maya\api\lib.py", line 2419, in reset_scene_resolution
#     asset_data = get_current_project_asset()["data"]
# TypeError: 'NoneType' object is not subscriptable
```

**Comment by iLLiCiTiT** [comment]
> Easiest way to test is to simply open maya and try to publish a workfile.

Should be fixed.

> in maya during startup - this is probably related:

This didn't happen to me. Need more information, like what `os.environ["AVALON_ASSET"]` contains or what `get_current_asset_name` returns.

**Comment by BigRoy** [comment]
> > Easiest way to test is to simply open maya and try to publish a workfile.
> 
> Should be fixed.
> 
> > in maya during startup - this is probably related:
> 
> This didn't happen to me. Need more information, like what `os.environ["AVALON_ASSET"]` contains or what `get_current_asset_name` returns.

@iLLiCiTiT I suspect the issue is [here](https://github.com/ynput/OpenPype/blob/f2e14e7f4aa4c9d1557e3fdacc05cc564c21d7bb/openpype/pipeline/context_tools.py#L436-L441)

Does a call to `get_current_project_asset()` for you return the asset data correctly (from within an asset with a nested path identifier?)

**Comment by antirotor** [comment]
Old instances (in Maya) are not working in the Publisher. `Asset` on instance remains and `Asset Path` points to the old asset:

image

**Comment by antirotor** [comment]
In blender (3.5), the asset path is used in paths - `extract_camera_abc.py` - is producing representation like:
```javascript
{
    'name': 'abc',
    'ext': 'abc',
    'files': '/assets/Alice_cameraMain.abc',
    'stagingDir': 'C:\\Users\\annat\\AppData\\Local\\Temp\\pyblish_tmp_5jrfkrxx'
}
```
this is for asset path `/assets/Alice`

**Comment by simonebarbieri** [comment]
In blender I'm getting the path in the asset:
image

It fails to extract then because it fails to join correctly the staging folder and the asset name:
image

**Comment by moonyuet** [comment]
So far so good in 3dsMax
Launcher
image
Loader
image
I even tried to switch Folder and Task in Publisher, it looks good also.

**Comment by simonebarbieri** [comment]
In unreal the folder has the full path:
image

But it publishes and loads it correctly:
image

**Comment by moonyuet** [comment]
The work folder setup looks good via workfile saving with different task type and folder type if the work folder already exists.
image
image
image
I double check the issue I spotted(saving the workfile from `lenny_bedroom/asset/art` to `lenny_restroom/asset/generic`, and I find this happened only when the folder path `asset/` doesn't exist yet in the work folder `lenny_restroom`. It only happens to the folder type but not the task type(the workfile type saving to the task type looks good when the work folder is already there). I guess this is expected behavior.

**Comment by MustafaJafar** [comment]
I faced these in Houdini 
image
image

**Comment by antirotor** [comment]
> I faced these in Houdini

this is because you have a scene with instances not using asset path but just asset - similar problem happens in Maya and elsewhere. Should be enough to fix by re-selecting asset. This is the same issue as https://github.com/ynput/OpenPype/pull/5817#issuecomment-1798240029

**Comment by MustafaJafar** [comment]
I did more testing in Houdini. 
1. It behaves as usual.
    image
2. I need to figure out how to update `Folder` value when creating publish instances (currently, I change them manually) 
3. workfile publish instance is updated automatically (open, close, open the publisher UI). i.e. it's not created with the correct value.
4. HDA publish instance was published successfully after updating `Folder` value but there are few other parameters to fix too . Here's how it looks like after loading. 
    image
5. for some reason I faced a problem with hero version integration. thus, I had to disable it. 
      ```
      DEBUG: --- Integration of Hero version for subset `pointcacheBgeo` begins.
      DEBUG: Looking for matching profile for: hosts: "houdini" | families: "pointcache" | task_names: "fx" | task_types: "FX"
      DEBUG: "houdini" not found in "hosts": ['standalonepublisher']
      DEBUG: None of profiles match your setup. hosts: "houdini" | families: "pointcache" | task_names: "fx" | task_types: "FX"
      DEBUG: `hero` template check was successful. `{root[work]}/{project[name]}/{hierarchy}/{folder[name]}/publish/{family}/{subset}/hero/{project[code]}_{folder[name]}_{task[name]}_hero<_{comment}>.{ext}`
      DEBUG: hero publish dir: "\\storage\work\ayon_projects\Test_Project\Factory_01\Machine_02\publish\pointcache\pointcacheBgeo\hero"
      DEBUG: http://192.168.1.4:5000 "POST /graphql HTTP/1.1" 200 105
      DEBUG: Response <RestApiResponse [200]>
      DEBUG: Creating first hero version.
      DEBUG: http://192.168.1.4:5000 "POST /graphql HTTP/1.1" 200 112
      DEBUG: Response <RestApiResponse [200]>
      ERROR: !!! Creating of hero version failed. Previous hero version maybe lost some data!
      Traceback (most recent call last):
        File "C:\Users\Mustafa Taher\AppData\Local\Ynput\AYON\dependency_packages\ayon_2308181924_windows.zip\dependencies\pyblish\plugin.py", line 527, in __explicit_process
          runner(*args)
        File "C:\Users\Mustafa Taher\AppData\Local\Ynput\AYON\addons\openpype_3.17.5\openpype\plugins\publish\integrate_hero_version.py", line 97, in process
          self.integrate_instance(
        File "C:\Users\Mustafa Taher\AppData\Local\Ynput\AYON\addons\openpype_3.17.5\openpype\plugins\publish\integrate_hero_version.py", line 364, in integrate_instance
          head, tail = _template_filled.split(frame_splitter)
      ValueError: not enough values to unpack (expected 2, got 1)
      ```

**Comment by iLLiCiTiT** [comment]
> 3. I need to figure out how to update Folder value when creating publish instances (currently, I change them manually)

You don't need to figure out anything. I'm working on it.

>  it's not created with the correct value.

Will look into it.

> but there are few other parameters to fix too

What other parameters?

> Here's how it looks like after loading.

That is loaded container, not publish instance.

> for some reason I faced a problem with hero version integration. thus, I had to disable it.

Unrelated to this PR, but deserver an issue.

**Comment by MustafaJafar** [comment]
> > but there are few other parameters to fix too
> 
> What other parameters?

These: 
image

**Comment by MustafaJafar** [comment]
I gave it another test in Houdini, 
I tested it many times and it works as expected (I can find published files where it should be).
sometimes I need to set `Folder` value in publisher UI manually, but as as I do that it works.

**Comment by iLLiCiTiT** [comment]
> Isn't that still a bug in itself? If so, when does that happen?

I do expect it happens only if the previous asset name is in project more than once (there is not "good" autofix). @MustafaJafar can you confirm?

**Comment by MustafaJafar** [comment]
> > Isn't that still a bug in itself? If so, when does that happen?
> 
> I do expect it happens only if the previous asset name is in project more than once (there is not "good" autofix). @MustafaJafar can you confirm?

I'm not sure what `if the previous asset name is in project more than once`  means. 

I tried opening different files, different assets and saving workfiles into other assets,
it seems that it works fine (speaking of `Folder` and `folderPath` especially workfile instance)

**Comment by iLLiCiTiT** [comment]
> I'm not sure what if the previous asset name is in project more than once means.

It means that if asset was `Machine_01`, it would stay as `Machine_01` (and be invalid) because it don't know which of the 2 it is.

> created instances (by tab menu) still require updating folderPath manually

I don't know what that is, could you please debug it and try to fix? Please create branch from this PR and create another PR.

**Comment by MustafaJafar** [comment]
> It means that if asset was Machine_01, it would stay as Machine_01 (and be invalid) because it don't know which of the 2 it is.

I could replicate it. 
image

but it's even more scary here 
image

**Comment by simonebarbieri** [comment]
Creators don't work in Blender after last merge.

Need to change this: https://github.com/ynput/OpenPype/blob/6fece8f6eacdc31a57bfaebe6b8075baa038ffea/openpype/hosts/blender/api/plugin.py#L228

to:
```
name = asset_name(instance_data["folderPath"], subset_name)
```

**Comment by mkolar** [comment]
Minor issue with the menu in nuke 14.1
image

Write creators are expeting `asset` in the data which is not being satisfied, hence Render, Image and Prerender createor fail
```
Creator: Render (write) (create_write_render)

Error: 'asset'

Traceback (most recent call last):
  File "/Users/milan/code/OpenPype/openpype/pipeline/create/context.py", line 2053, in _create_with_unified_error
    result = creator.create(*args, **kwargs)
  File "/Users/milan/code/OpenPype/openpype/hosts/nuke/plugins/create/create_write_render.py", line 93, in create
    instance_node = self.create_instance_node(
  File "/Users/milan/code/OpenPype/openpype/hosts/nuke/plugins/create/create_write_render.py", line 62, in create_instance_node
    created_node = napi.create_write_node(
  File "/Users/milan/code/OpenPype/openpype/hosts/nuke/api/lib.py", line 1292, in create_write_node
    anatomy_filled = format_anatomy(data)
  File "/Users/milan/code/OpenPype/openpype/hosts/nuke/api/lib.py", line 1129, in format_anatomy
    asset_name = data["asset"]
KeyError: 'asset'
```

**Comment by iLLiCiTiT** [comment]
> Minor issue with the menu in nuke 14.1

Should be fixed with https://github.com/ynput/OpenPype/pull/5887

**Comment by simonebarbieri** [comment]
In Blender, I'm getting this error when loading a layout with a rig:
```
Traceback (most recent call last):
  File "C:\Dev\Ynput\OpenPype\openpype\hosts\blender\api\ops.py", line 113, in execute
    result = callback(*args, **kwargs)
  File "C:\Dev\Ynput\OpenPype\openpype\hosts\blender\api\plugin.py", line 476, in _load
    nodes = self.process_asset(
  File "C:\Dev\Ynput\OpenPype\openpype\hosts\blender\plugins\load\load_blend.py", line 149, in process_asset
    self._post_process_layout(container, asset, representation)
  File "C:\Dev\Ynput\OpenPype\openpype\hosts\blender\plugins\load\load_blend.py", line 69, in _post_process_layout
    create_context.create(
  File "C:\Dev\Ynput\OpenPype\openpype\pipeline\create\context.py", line 2029, in create
    return creator.create(
  File "C:\Dev\Ynput\OpenPype\openpype\hosts\blender\plugins\create\create_animation.py", line 18, in create
    collection = super().create(
  File "C:\Dev\Ynput\OpenPype\openpype\hosts\blender\api\plugin.py", line 229, in create
    asset_name = instance_data["folderPath"]
KeyError: 'folderPath'
```

It happens because when loading a layout with a rig, it will create an animation instance for each rig:
```
        # Create animation instances for each rig
        creator_identifier = "io.openpype.creators.blender.animation"
        host = registered_host()
        create_context = CreateContext(host)

        for rig in rigs:
            create_context.create(
                creator_identifier=creator_identifier,
                variant=rig.name.split(':')[-1],
                pre_create_data={
                    "use_selection": False,
                    "asset_group": rig
                }
            )
```

But when using the `create` from `CreateContext`, `instance_data` is defined as:
https://github.com/ynput/OpenPype/blob/79c079961bac7ddffa1351d31f9e9fa370f24909/openpype/pipeline/create/context.py#L2023-L2028

It's missing `folderPath` if `AYON_SERVER_ENABLED`.

**Comment by iLLiCiTiT** [comment]
> In Blender, I'm getting this error when loading a layout with a rig:

Fixed.

**Comment by simonebarbieri** [comment]
After latest fix, everything seems to be working fine in Blender :)

**Review by LiborBatek — APPROVED** [review]
I have made tests in SubstancePainter and all working fine with paths as unique identifier.

All good.

**Review by mkolar — APPROVED** [review]

**Pull request merged**
